### PR TITLE
Optimize SpanBuilderSdk link memory

### LINKED ISSUE #709: SpanBuilderSdk.truncatedLinks may keep references to thrown away links
`truncatedLinks` uses `ArrayList.subList` to truncate the links.

https://github.com/open-telemetry/opentelemetry-java/blob/7505c1297f7eaf324b16a68d07142a93a5472b50/sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java#L226

However, `subList` returns a `ArrayList.SubList` object which has a reference to the full parent list (along with the start/end index to which it refers). Since the purpose of `trucatedLinks` seems to be to bound memory consumption, the method seems to fail to fulfill its purpose.

I suggest removing `truncatedLinks` and doing the check right in `addLink`.

### PR DESCRIPTION
**Pull request by jkwatson**
resolves #709

### CODE REVIEW

**Comment by carlosalberto** [comment]
This looks fine to me - and while on that subject, I've long wondered if we should keep the same behavior we have with `Attribute`s, dropping the eldest entry once we have reached the maximum allowed count.

Maybe @bogdandrutu remembers about this, since we are using OC based code? ;)

**Comment by jkwatson** [comment]
Do you know why we would drop the "eldest" entry, rather than just dropping any that exceed the maximum? Is there a reason why recency would mean more relevance?

File: sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java [hunk 1]
```diff
@@ -127,10 +128,18 @@
   @Override
   public Span.Builder addLink(Link link) {
     Utils.checkNotNull(link, "link");
+    totalNumberOfLinksAdded++;
+    // don't bother doing anything with any links beyond the max.
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java]: It is better to drop the earliest in general.

File: sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java [hunk 2]
```diff
@@ -127,10 +128,18 @@
...
+    // todo: generate a metric for dropped links
```
Comment (bogdandrutu) [inline sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java]: I think this makes more sense to be an issue for all dropped things (attributes, links, events).

File: sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java (hunk 1 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java]: Can you explain why? Are later links somehow better than earlier ones? What's the business rationale here? [the code is much more efficient if we can just drop the later ones]

File: sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java (hunk 2 above)
Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java]: definitely

Comment (bogdandrutu) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java]: so suggestion is to remove the todo and create an issue to discuss and make a decision on what to do with the dropped things.

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java]: ok, will do.

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java]: https://github.com/open-telemetry/opentelemetry-java/issues/766

Comment (jkwatson) [reply sdk/src/main/java/io/opentelemetry/sdk/trace/SpanBuilderSdk.java]: todo removed.

**Pull request merged**